File: src/core/semantics/code/interpreter.rb
#=begin
#
# Some definitions for documentation below:
# - operation: an action the interpreter can take, eg eval, debug, lvalue, etc
# - type: schema type, eg BinaryOp
# - method: name of method for {o:operation X t:type} eg eval_BinaryOp()
#=end

require 'enso'

module Interpreter
  class DynamicPropertyStack
    def initialize
      @current = {}
      @stack = []
    end
    
    def [](name)
      @current[name]
    end

    def include?(name)
      @current.include?(name)
    end
    
    def keys
      @current.keys
    end
    
    def _bind(field, value)
      if @current.has_key?(field)
        old = @current[field]
      else
        old = :undefined
      end
      @stack << [field, old]
      @current[field] = value
    end
    
    def _pop(n = 1)
      while (n > 0) do
        parts = @stack.pop
        if parts[1] == :undefined
          @current.delete(parts[0])
        else
          @current[parts[0]] = parts[1]
        end
        n -= 1
      end
    end

    def to_s
      @current.to_s
    end
  end
  
  module Dispatcher
    def init
      if !@D
        @D = DynamicPropertyStack.new
      end
      @indent = nil
    end

    def dynamic_bind(fields={}, &block)
      if !@D
        @D = DynamicPropertyStack.new
      end
      fields.each do |key, value|
        @D._bind(key, value)
      end
      result = block.call
      @D._pop(fields.size)
      result
    end

    def wrap(operation, outer, obj)
      init_done = @init
      if !init_done
        init
      end
      @init = true
      type = obj.schema_class
      method = "#{outer}_#{type.name}".to_s
      if !respond_to?(method)
        method = find_op(outer, type)  # slow path
      end
      if !method
        method = "#{outer}_?".to_s
        if !respond_to?(method)
          raise "Missing method in interpreter for #{outer}_#{type.name}(#{obj})"
        end
      end
      result = nil
      send(method, obj) {
        result = dispatch_obj(operation, obj)
      }
      if !init_done
        @init = false
      end
      result
    end

    def dispatch_obj(operation, obj)
      init_done = @init
      if !init_done
        init
      end
      @init = true
      type = obj.schema_class
      method = "#{operation}_#{type.name}".to_s
      if !respond_to?(method)
        method = find_op(operation, type)  # slow path
      end
      if !method
        method = "#{operation}_?".to_s
        if !respond_to?(method)
          raise "Missing method in interpreter for #{operation}_#{type.name}(#{obj})"
        end
      end
      if @indent
      	STDERR.puts "#{' '*@indent}#{method}"
      	@indent = @indent + 1
     	end
      result = send(method, obj)
      if @indent
      	STDERR.puts "#{' '*@indent}=#{result}"
      	@indent = @indent - 1
     	end
      if !init_done
        @init = false
      end
      result
    end

    def find_op(operation, type)
      method = "#{operation}_#{type.name}".to_s
      if respond_to?(method)
        method
      else
        type.supers.find_first do |p| 
          find_op(operation, p)
        end
      end
    end
  end
end

File: src/js/core/semantics/code/interpreter.js
'use strict'

//// Interpreter ////

var cwd = process.cwd() + '/';
var Enso = require(cwd + "enso.js");

var Interpreter;

class DynamicPropertyStack {
  static new(...args) { return new DynamicPropertyStack(...args) };

  constructor() {
    var self = this;
    self.current$ = Enso.EMap.new();
    self.stack$ = [];
  };

  get$(name) {
    var self = this;
    return self.current$.get$(name);
  };

  include_P(name) {
    var self = this;
    return self.current$.include_P(name);
  };

  keys() {
    var self = this;
    return self.current$.keys();
  };

  _bind(field, value) {
    var self = this, old;
    if (self.current$.has_key_P(field)) {
      old = self.current$.get$(field);
    } else {
      old = "undefined";
    }
    self.stack$.push([field, old]);
    return self.current$ .set$(field, value);
  };

  _pop(n = 1) {
    var self = this, parts;
    while (n > 0) {
      parts = self.stack$.pop();
      if (parts.get$(1) == "undefined") {
        self.current$.delete_M(parts.get$(0));
      } else {
        self.current$ .set$(parts.get$(0), parts.get$(1));
      }
      n = n - 1;
    }
  };

  to_s() {
    var self = this;
    return self.current$.to_s();
  };
};

function Dispatcher(parent) {
  return class extends parent {
    init() {
      var self = this;
      if (! self.D$) {
        self.D$ = DynamicPropertyStack.new();
      }
      return self.indent$ = null;
    };

    dynamic_bind(block, fields = Enso.EMap.new()) {
      var self = this, result;
      if (! self.D$) {
        self.D$ = DynamicPropertyStack.new();
      }
      fields.each(function(key, value) {
        return self.D$._bind(key, value);
      });
      result = block();
      self.D$._pop(fields.size_M());
      return result;
    };

    wrap(operation, outer, obj) {
      var self = this, init_done, type, method, result;
      init_done = self.init$;
      if (! init_done) {
        self.init();
      }
      self.init$ = true;
      type = obj.schema_class();
      method = Enso.S(outer, "_", type.name()).to_s();
      if (! self.respond_to_P(method)) {
        method = self.find_op(outer, type);
      }
      if (! method) {
        method = Enso.S(outer, "_?").to_s();
        if (! self.respond_to_P(method)) {
          self.raise(Enso.S("Missing method in interpreter for ", outer, "_", type.name(), "(", obj, ")"));
        }
      }
      result = null;
      self.send(function() {
        return result = self.dispatch_obj(operation, obj);
      }, method, obj);
      if (! init_done) {
        self.init$ = false;
      }
      return result;
    };

    dispatch_obj(operation, obj) {
      var self = this, init_done, type, method, result;
      init_done = self.init$;
      if (! init_done) {
        self.init();
      }
      self.init$ = true;
      type = obj.schema_class();
      method = Enso.S(operation, "_", type.name()).to_s();
      if (! self.respond_to_P(method)) {
        method = self.find_op(operation, type);
      }
      if (! method) {
        method = Enso.S(operation, "_?").to_s();
        if (! self.respond_to_P(method)) {
          self.raise(Enso.S("Missing method in interpreter for ", operation, "_", type.name(), "(", obj, ")"));
        }
      }
      if (self.indent$) {
        STDERR.puts(Enso.S(" " * self.indent$, method));
        self.indent$ = self.indent$ + 1;
      }
      result = self.send(method, obj);
      if (self.indent$) {
        STDERR.puts(Enso.S(" " * self.indent$, "=", result));
        self.indent$ = self.indent$ - 1;
      }
      if (! init_done) {
        self.init$ = false;
      }
      return result;
    };

    find_op(operation, type) {
      var self = this, method;
      method = Enso.S(operation, "_", type.name()).to_s();
      if (self.respond_to_P(method)) {
        return method;
      } else {
        return type.supers().find_first(function(p) {
          return self.find_op(operation, p);
        });
      }
    }; }};

Interpreter = {
  DynamicPropertyStack: DynamicPropertyStack,
  Dispatcher: Dispatcher,
};
module.exports = Interpreter ;
